### ISSUE TITLE
multiple parallel Wake invocations

### ISSUE THREAD

**Issue by albertchen-sifive** [sifive/wake#403]
It would be really really nice if I could have more than one wake build running at the same time in the same workspace. something like a wake build server

some use cases I'm thinking of
- I've already started a long-running job and would like to build other stuff in the meantime.
- I have a program that wants to use existing wake rules to build many things, inspecting intermediate outputs to determine what to build next
  - CI script that incrementally builds multiple things using wake and checks the output files at each step
  - wrapper script that presents a nicer CLI
  - GUIs, IDEs

**Comment by terpstra** [comment]
Yep. Unless I run into some unexpected hurdle, this going to be one of the features in 0.19.

**Comment by terpstra** [comment]
BTW, the way it will work is like this:
- First wake invocation starts running normally
- Second wake invocation with no special args goes into the 'waiting for my turn' loop we have now, but prints a new help message
- Second wake invocation with `--no-wait` fails immediately and prints the new help message
- Second wake invocation with `--join` begins running in the shared build

The new message is this:

There is already a wake build running in this workspace. You can add more tasks to this build by invoking wake with `--join`. However, the builds will be combined, which means that any work already completed by the currently running build will not be re-executed by your new tasks. Therefore, you should only use `--join` if you have made no changes to the workspace since the initial build was started. (Really, you should not change the workspace during a normal build anyway.)

**Comment by terpstra** [comment]
For your CI example, though, I don't see how wake's current behaviour is a problem. You run a 'wake step1', inspect the output, then run 'wake step2' (which reruns 'step1', but thanks to the magic of incremental builds, this is nearly instant), repeat.

**Comment by albertchen-sifive** [comment]
great thanks. for the CI example I was thinking of the case where you have a large suite of tests that are all running incrementally and in parallel.

**Comment by terpstra** [comment]
Right. That scenario should work with this future `--join` functionality.

**Comment by albertchen-sifive** [comment]
Say a script does `wake -x 'write "foo" "foo.txt"'` followed by `wake --join -x 'write "bar "foo.txt"'`. if the first job finishes before the second wake invocation starts, then it would not catch the `file foo.txt output by multiple jobs`, whereas if the second one is able to join the first job in time then it would catch the error.  Am i understanding this correctly?

I think usually scripts would want to make sure all invocations are joined together even if they don't overlap. Could there be something like a `--persistent` flag that then makes all following wake invocations part of the same build.

**Comment by terpstra** [comment]
That seems reasonable. We could have explicit '--start' and '--stop' control. BTW, there will be a technical restriction that you can't use '--join' with '-x'. You will only be able to request target execution (the shared build will share the compiled wake sources).

**Comment by terpstra** [comment]
So this did not make it into 0.19, but it's still on my radar for when the rule-action rework is implemented.
